### PR TITLE
Feature/whitelist

### PR DESCRIPTION
**Pull request by ysiad**
Pivotal Tracker Stories:

https://www.pivotaltracker.com/story/show/132354667
https://www.pivotaltracker.com/story/show/132354743

Added whitelist feature, including:
1. whitelist view
2. Add user to whitelist
3. Delete user from whitelist

### CODE REVIEW

File: db/migrate/20160720004301_create_basic_tables.rb
Comment (junyu-w) [inline db/migrate/20160720004301_create_basic_tables.rb]: do not modify existing migration because changes won't be reflected to developers who has already run this migration. Instead create table by generating a new migration

File: db/migrate/20161020040440_create_root_user.rb
Comment (junyu-w) [inline db/migrate/20161020040440_create_root_user.rb]: no need to add `self.`, same for `self.down`

File: db/migrate/20161021113555_create_normal_user.rb
Comment (junyu-w) [inline db/migrate/20161021113555_create_normal_user.rb]: don't need migration to create `coach` user. because `coach` should only be created through external login from github.

File: db/migrate/20160720004301_create_basic_tables.rb
Comment (junyu-w) [inline db/migrate/20160720004301_create_basic_tables.rb]: `role` is a duplicated column for `authorized_users` table since it's already in `users`. If this is absolutely needed a better way is to create a `roles` table and associate it with both `authorized_users` an `users`. I would recommend delete this column.

File: app/models/authorized_user.rb
Comment (junyu-w) [inline app/models/authorized_user.rb]: model name should be CamelCase, and should be the same as corresponding controller name (controller class name is the plural form of model name) so this should be `AuthorizedUser`

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: Controller name should be `AuthorizedUsersController`

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: this part (line 45-52) won't be needed if `role` column is removed

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: email format checking can be replaced with this one line at the very top using built in rails validation 
`validates_format_of :email,:with => Devise::email_regexp`

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: hmm show should be corresponding to some particular whitelist? why does this part redirect to /whitelist/index?

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: `respond_to....` can be replaced with code in similar fashion as the else block? is there a specific reason to specify format here?

File: app/models/authorized_user.rb
Comment (junyu-w) [inline app/models/authorized_user.rb]: `find_by` is not a method of a particular model object, it's a class method
so this function can be replaced with a one liner `!AuthorizedUser.find_by_email(email).nil?`

File: app/models/user.rb
Comment (junyu-w) [inline app/models/user.rb]: `puts` should be removed from PR :)

File: app/views/whitelist/_add.html.haml
Comment (junyu-w) [inline app/views/whitelist/_add.html.haml]: better use rails route helper like `whitelist_path` instead of hard coding url

File: db/migrate/20161021104336_create_authorized_user.rb
Comment (junyu-w) [inline db/migrate/20161021104336_create_authorized_user.rb]: This migration is not needed because only the first admin needs to be created through migration.

**Review by junyu-w — CHANGES_REQUESTED** [review]
some changes need to be made before merging 👍

File: db/migrate/20161021113555_create_normal_user.rb
Comment (ysiad) [reply db/migrate/20161021113555_create_normal_user.rb]: Fixed.

File: app/controllers/whitelist_controller.rb
Comment (ysiad) [reply app/controllers/whitelist_controller.rb]: Don't see why. ..Did not implement this in the new commit

File: db/migrate/20161021104336_create_authorized_user.rb
Comment (ysiad) [reply db/migrate/20161021104336_create_authorized_user.rb]: Fixed.

Comment (ysiad) [reply db/migrate/20161021104336_create_authorized_user.rb]: Fixed all the issues except the controller name issue as the comments suggested. Too lazy to reply other comments.

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [reply app/controllers/whitelist_controller.rb]: It's a rails convention that controller name should be the plural format of model name, and since rails is a 'convention over configuration' framework most people will assume controller name is "AuthorizedUsersController" if model name is "AuthorizedUser"

File: app/models/authorized_user.rb
Comment (junyu-w) [inline app/models/authorized_user.rb]: hmm does this `has_email?` function work? shouldn't it be `AuthorizedUser.find_by_email(...)`

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: the email regex check using `validate` is missing..

Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: Since all controller actions requires admin permission and the else blocks are all the same, it's better to define a function like 

```
def check_if_admin
  unless current_user.is_admin?
     flash[:notice] = ...
     redirect_to .... and return
  end
end
```

and add this line at the top
`before_action :check_if_admin`

**Review by junyu-w — CHANGES_REQUESTED** [review]
added some more suggestions

File: app/controllers/whitelist_controller.rb
Comment (ysiad) [reply app/controllers/whitelist_controller.rb]: Fixed.

File: app/controllers/whitelist_controller.rb
Comment (ysiad) [reply app/controllers/whitelist_controller.rb]: Changed the model name to whitelist

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: is this line needed in controller?

File: app/models/whitelist.rb
Comment (junyu-w) [inline app/models/whitelist.rb]: you can run `annotate` in command line and this will be automatically updated

**Review by junyu-w — COMMENTED** [review]
good to go after some minor changes 👍  + merge latest commits from feature/login-with-github

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [inline app/controllers/whitelist_controller.rb]: maybe you mean `projects_path` here?

File: app/controllers/whitelist_controller.rb
Comment (junyu-w) [reply app/controllers/whitelist_controller.rb]: Controller name should be `WhitelistsController`

File: app/models/whitelist.rb
Comment (junyu-w) [inline app/models/whitelist.rb:12]: just `!Whitelist.find_by_email(email).nil?` is good

File: config/routes.rb
Comment (junyu-w) [inline config/routes.rb]: `resources :whitelists` after controller name change

**Review by junyu-w — CHANGES_REQUESTED** [review]

**Review by junyu-w — APPROVED** [review]

**Pull request closed without merging**
